tests/e2e/apps/hop/app.go

```diff
@@ -0,0 +1,253 @@
+	clientTimeout time.Duration
+}
+
+func (a App) makeHTTPRequest(req *config.HopMessage, url string) (*config.HopMessage, error) {
+	glog.Infof("Making HTTP Request to %s", url)
+	jsonStr, err := a.marshaller.MarshalToString(req)
+	if err != nil {
+		return nil, err
+	}
+	hreq, err := http.NewRequest("POST", url, bytes.NewBuffer([]byte(jsonStr)))
+	if err != nil {
+		return nil, err
+	}
+	hreq.Header.Set("X-Custom-Header", "myvalue")
+	hreq.Header.Set("Content-Type", "application/json")
+	client := &http.Client{}
+	resp, err := client.Do(hreq)
+	defer func() {
+		if e := resp.Body.Close(); e != nil {
+			glog.Error(err)
+		}
+	}()
+	if err != nil {
+		return nil, err
+	}
+	pb := new(config.HopMessage)
+	if err = a.unmarshaller.Unmarshal(resp.Body, pb); err != nil {
+		return nil, err
+	}
+	return pb, err
+}
+
+func (a App) makeGRPCRequest(req *config.HopMessage, address string) (*config.HopMessage, error) {
+	glog.Infof("Making GRPC Request to %s", address)
+	conn, err := grpc.Dial(address,
+		grpc.WithInsecure(),
+		// grpc-go sets incorrect authority header
+		grpc.WithAuthority(address),
+		grpc.WithBlock(),
+		grpc.WithTimeout(a.clientTimeout))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if e := conn.Close(); e != nil {
+			glog.Error(e)
+		}
+	}()
+	client := config.NewHopTestServiceClient(conn)
+	resp, err := client.Hop(context.Background(), req)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// ServerHTTP starts a HTTP Server for Hop App
+func (a App) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	req := new(config.HopMessage)
+	err := a.unmarshaller.Unmarshal(r.Body, req)
+	if err != nil {
+		glog.Error(err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	p := req.GetPosition()
+	glog.Infof("HTTP Serving message %s at position %d", req.GetId(), p)
+	resp := a.forwardMessage(req)
+	jsonStr, err := a.marshaller.MarshalToString(resp)
+	if err != nil {
+		glog.Error(err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+	_, err = w.Write([]byte(jsonStr))
+	if err != nil {
+		glog.Error(err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	glog.Infof("Successfully served message %s at position %d", req.GetId(), p)
+}
+
+// Hop start a gRPC server for Hop App
+func (a App) Hop(ctx context.Context, req *config.HopMessage) (*config.HopMessage, error) {
+	p := req.GetPosition()
+	glog.Infof("GRPC Serving message %s at position %d", req.GetId(), p)
+	resp := a.forwardMessage(req)
+	glog.Infof("Successfully served message %s at position %d", req.GetId(), p)
+	return resp, nil
+}
+
+func (a App) makeRequest(m *config.HopMessage, d string) (*config.HopMessage, error) {
+	// Check destination and send using grpc or http handler
+	if strings.HasPrefix(d, "http://") {
+		return a.makeHTTPRequest(m, d)
+
+	} else if strings.HasPrefix(d, "grpc://") {
+		address := d[len("grpc://"):]
```

Comment:
Use strings.TrimPrefix. It might also be more idiomatic to rewrite with an expressionless switch, i.e.

```golang
switch {
case strings.HasPrefix(d, "http://"):
    return a.makeHTTPRequest(m, d)
case strings.HasPrefix(d, "grpc://"):
    return a.makeGRPCRequest(m, strings.TrimPrefix(d, "grpc://"))
default:
    return nil, errors.New("protocol not supported")
}

```